grakn-graph/src/main/java/ai/grakn/graph/internal/ElementFactory.java

```diff
@@ -131,89 +134,116 @@ RoleTypeImpl buildRoleType(Vertex vertex, RoleType type, Boolean isImplicit){
     }
 
-    //TODO: Simplify this if it does not make a difference to large loading
-    private Schema.BaseType getBaseType(Vertex vertex){
+    /**
+     * This is a helper method to get the base type of a vertex.
+     * It first tried to get the base type via the label.
+     * If this is not possible it then tries to get the base type via the Shard Edge.
+     *
+     * @param vertex The vertex to build a concept from
+     * @return The base type of the vertex, if it is a valid concept.
+     */
+    private Schema.BaseType getBaseType(VertexElement vertex){
         try {
             return Schema.BaseType.valueOf(vertex.label());
         } catch (IllegalArgumentException e){
-            //Base type appears to be invalid. Let's try getting the type via the isa edge
-            Iterator<Edge> iterator = vertex.edges(Direction.OUT, Schema.EdgeLabel.SHARD.getLabel());
-            if(iterator.hasNext()){
-                Vertex shardVertex = iterator.next().inVertex();
-                Vertex typeVertex = shardVertex.edges(Direction.OUT, Schema.EdgeLabel.ISA.getLabel()).next().inVertex();
-                String label = typeVertex.label();
+            //Base type appears to be invalid. Let's try getting the type via the shard edge
+            Optional<EdgeElement> type = vertex.getEdgesOfType(Direction.OUT, Schema.EdgeLabel.SHARD).findAny();
+
+            if(type.isPresent()){
+                String label = type.get().target().label();
                 if(label.equals(Schema.BaseType.ENTITY_TYPE.name())) return Schema.BaseType.ENTITY;
-                if(label.equals(Schema.BaseType.RELATION_TYPE.name())) return Schema.BaseType.RELATION;
+                if(label.equals(RELATION_TYPE.name())) return Schema.BaseType.RELATION;
                 if(label.equals(Schema.BaseType.RESOURCE_TYPE.name())) return Schema.BaseType.RESOURCE;
-                if(label.equals(Schema.BaseType.RULE_TYPE.name())) return Schema.BaseType.RULE;
+                if(label.equals(RULE_TYPE.name())) return Schema.BaseType.RULE;
             }
         }
-        throw new IllegalStateException("Could not determine the base type of vertex [" + vertex + "]");
+        throw new IllegalStateException("Could not determine the base type of vertex [" + vertex.element() + "]");
     }
 
+    // ---------------------------------------- Non Concept Construction -----------------------------------------------
     EdgeElement buildEdge(Edge edge){
         return new EdgeElement(graknGraph, edge);
     }
 
-    //TODO: Integrate with cache and cleanup
     Casting buildRolePlayer(Edge edge){
         return buildRolePlayer(buildEdge(edge));
     }
 
     Casting buildRolePlayer(EdgeElement edge) {
-        return new Casting(graknGraph, edge.getEdge());
+        return new Casting(edge);
+    }
+
+    Shard buildShard(ConceptImpl shardOwner, Vertex vertex){
+        return new Shard(shardOwner, buildVertexElement(vertex));
```

Comment:
Some methods here take `Vertex`, some `VertexElement`. It feels like they should all take `VertexElement` to me.